File: CodeNugde/CodeNugde/src/app/services/ai-evaluation.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ApiService } from './api.service';

// AI Evaluation interfaces
export interface CodeSubmission {
  questionId: string;
  code: string;
  language: string;
  userId: string;
}

export interface AIEvaluationResult {
  score: number; // 0-100
  feedback: string;
  suggestions: string[];
  codeQuality: CodeQualityMetrics;
  testResults: TestResult[];
  overallAssessment: string;
}

export interface CodeQualityMetrics {
  readability: number; // 0-10
  efficiency: number; // 0-10
  maintainability: number; // 0-10
  bestPractices: number; // 0-10
}

export interface TestResult {
  testCase: string;
  passed: boolean;
  expected: string;
  actual: string;
  executionTime: number;
}

export interface AIFeedbackRequest {
  code: string;
  language: string;
  questionDescription: string;
  expectedOutput?: string;
  testCases?: TestCase[];
}

export interface TestCase {
  input: string;
  expectedOutput: string;
  description: string;
}

@Injectable({
  providedIn: 'root'
})
export class AIEvaluationService {
  private readonly apiService: ApiService;

  constructor(
    private http: HttpClient,
    apiService: ApiService
  ) {
    this.apiService = apiService;
  }

  // Submit code for AI evaluation
  evaluateCode(submission: CodeSubmission): Observable<AIEvaluationResult> {
    return this.apiService.post<AIEvaluationResult>('/AI/evaluate', submission);
  }

  // Get AI feedback for code improvement
  getCodeFeedback(request: AIFeedbackRequest): Observable<string> {
    return this.apiService.post<string>('/AI/feedback', request);
  }

  // Get code suggestions for optimization
  getOptimizationSuggestions(code: string, language: string): Observable<string[]> {
    return this.apiService.post<string[]>('/AI/optimize', { code, language });
  }

  // Analyze code complexity
  analyzeComplexity(code: string, language: string): Observable<{
    timeComplexity: string;
    spaceComplexity: string;
    explanation: string;
  }> {
    return this.apiService.post('/AI/complexity', { code, language });
  }

  // Generate test cases for given code
  generateTestCases(code: string, language: string, description: string): Observable<TestCase[]> {
    return this.apiService.post<TestCase[]>('/AI/generate-tests', {
      code,
      language,
      description
    });
  }

  // Check for code plagiarism
  checkPlagiarism(code: string, questionId: string): Observable<{
    similarityScore: number;
    matches: Array<{
      userId: string;
      similarity: number;
      matchedLines: number[];
    }>;
  }> {
    return this.apiService.post('/AI/plagiarism', { code, questionId });
  }

  // Get code explanation
  explainCode(code: string, language: string): Observable<{
    explanation: string;
    keyPoints: string[];
    algorithmUsed?: string;
  }> {
    return this.apiService.post('/AI/explain', { code, language });
  }

  // Mock AI evaluation for development/testing
  mockEvaluateCode(submission: CodeSubmission): Observable<AIEvaluationResult> {
    return new Observable(observer => {
      setTimeout(() => {
        const mockResult: AIEvaluationResult = {
          score: Math.floor(Math.random() * 40) + 60, // 60-100
          feedback: this.generateMockFeedback(submission.code),
          suggestions: this.generateMockSuggestions(),
          codeQuality: {
            readability: Math.floor(Math.random() * 3) + 7, // 7-10
            efficiency: Math.floor(Math.random() * 4) + 6, // 6-10
            maintainability: Math.floor(Math.random() * 3) + 7, // 7-10
            bestPractices: Math.floor(Math.random() * 4) + 6 // 6-10
          },
          testResults: this.generateMockTestResults(),
          overallAssessment: 'Good solution with room for improvement in efficiency.'
        };
        observer.next(mockResult);
        observer.complete();
      }, 2000); // Simulate API delay
    });
  }

  private generateMockFeedback(code: string): string {
    const feedbacks = [
      'Your solution correctly implements the algorithm. Consider optimizing the time complexity.',
      'Good use of data structures. The code is readable and well-structured.',
      'The logic is correct but could be more efficient. Try using a different approach.',
      'Excellent solution! Clean code with optimal time and space complexity.',
      'The solution works but has some edge cases that need to be handled.'
    ];
    return feedbacks[Math.floor(Math.random() * feedbacks.length)];
  }

  private generateMockSuggestions(): string[] {
    const suggestions = [
      'Use more descriptive variable names',
      'Consider using a hash map for O(1) lookups',
      'Add input validation for edge cases',
      'Break down the function into smaller, reusable functions',
      'Add comments to explain complex logic',
      'Consider using built-in library functions',
      'Optimize the nested loops to reduce time complexity'
    ];
    
    // Return 2-4 random suggestions
    const count = Math.floor(Math.random() * 3) + 2;
    const shuffled = suggestions.sort(() => 0.5 - Math.random());
    return shuffled.slice(0, count);
  }

  private generateMockTestResults(): TestResult[] {
    const testCases = [
      'Basic functionality test',
      'Edge case: empty input',
      'Edge case: single element',
      'Large input test',
      'Negative numbers test'
    ];

    return testCases.map((testCase, index) => ({
      testCase,
      passed: Math.random() > 0.2, // 80% pass rate
      expected: `Expected output ${index + 1}`,
      actual: `Actual output ${index + 1}`,
      executionTime: Math.floor(Math.random() * 100) + 10 // 10-110ms
    }));
  }
}

File: BackEnd/codeN/CodeNudge/CodeNudge_backend/Infrastructure/Services/AIEvaluationService.cs
using CodeNudge.Core.Application.Interfaces;
using CodeNudge.Shared.Requests.AI;
using CodeNudge.Shared.Responses.AI;
using Microsoft.Extensions.Logging;
using System.Text.Json;
using AIEvaluationResult = CodeNudge.Shared.Responses.AI.AIEvaluationResult;

namespace CodeNudge.Infrastructure.Services;

public class AIEvaluationService : IAIEvaluationService
{
    private readonly ILogger<AIEvaluationService> _logger;
    private readonly HttpClient _httpClient;

    public AIEvaluationService(ILogger<AIEvaluationService> logger, HttpClient httpClient)
    {
        _logger = logger;
        _httpClient = httpClient;
    }

    public async Task<AIEvaluationResult> EvaluateCodeAsync(CodeEvaluationRequest request)
    {
        try
        {
            _logger.LogInformation("Evaluating code for user {UserId} and question {QuestionId}", 
                request.UserId, request.QuestionId);

            // For now, return a mock evaluation result
            // In production, this would call an AI service like OpenAI, Claude, or a custom ML model
            return await GenerateMockEvaluationAsync(request);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error evaluating code");
            throw;
        }
    }

    public async Task<string> GetCodeFeedbackAsync(CodeFeedbackRequest request)
    {
        try
        {
            _logger.LogInformation("Generating feedback for code in {Language}", request.Language);

            // Mock feedback generation
            await Task.Delay(1000); // Simulate AI processing time

            var feedbacks = new[]
            {
                "Your solution correctly implements the algorithm. Consider optimizing the time complexity by using a hash map for O(1) lookups instead of nested loops.",
                "Good use of data structures. The code is readable and well-structured. You might want to add input validation for edge cases.",
                "The logic is correct but could be more efficient. Try using a two-pointer approach to reduce the time complexity from O(n²) to O(n).",
                "Excellent solution! Clean code with optimal time and space complexity. The variable names are descriptive and the logic is easy to follow.",
                "The solution works but has some edge cases that need to be handled. Consider what happens with empty inputs or null values."
            };

            return feedbacks[new Random().Next(feedbacks.Length)];
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error generating code feedback");
            throw;
        }
    }

    public async Task<List<string>> GetOptimizationSuggestionsAsync(CodeOptimizationRequest request)
    {
        try
        {
            _logger.LogInformation("Generating optimization suggestions for {Language} code", request.Language);

            await Task.Delay(800); // Simulate processing time

            var allSuggestions = new[]
            {
                "Use more descriptive variable names to improve code readability",
                "Consider using a hash map for O(1) lookups instead of linear search",
                "Add input validation for edge cases (null, empty, boundary values)",
                "Break down the function into smaller, reusable functions",
                "Add comments to explain complex logic and algorithms",
                "Consider using built-in library functions for common operations",
                "Optimize nested loops to reduce time complexity",
                "Use early returns to reduce nesting and improve readability",
                "Consider memory usage - can you reduce space complexity?",
                "Add error handling for potential runtime exceptions"
            };

            // Return 3-5 random suggestions
            var random = new Random();
            var count = random.Next(3, 6);
            return allSuggestions.OrderBy(x => random.Next()).Take(count).ToList();
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error generating optimization suggestions");
            throw;
        }
    }

    public async Task<ComplexityAnalysisResult> AnalyzeComplexityAsync(CodeComplexityRequest request)
    {
        try
        {
            _logger.LogInformation("Analyzing complexity for {Language} code", request.Language);

            await Task.Delay(1200); // Simulate analysis time

            var complexities = new[]
            {
                new { Time = "O(n)", Space = "O(1)", Explanation = "Linear time complexity with constant space usage. The algorithm processes each element once." },
                new { Time = "O(n²)", Space = "O(1)", Explanation = "Quadratic time complexity due to nested loops. Space complexity is constant as no additional data structures are used." },
                new { Time = "O(n log n)", Space = "O(n)", Explanation = "Logarithmic time complexity, typically from sorting or divide-and-conquer approach. Linear space for auxiliary data structures." },
                new { Time = "O(1)", Space = "O(1)", Explanation = "Constant time and space complexity. The algorithm performs a fixed number of operations regardless of input size." }
            };

            var selected = complexities[new Random().Next(complexities.Length)];

            return new ComplexityAnalysisResult
            {
                TimeComplexity = selected.Time,
                SpaceComplexity = selected.Space,
                Explanation = selected.Explanation
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error analyzing code complexity");
            throw;
        }
    }

    public async Task<List<GeneratedTestCase>> GenerateTestCasesAsync(TestCaseGenerationRequest request)
    {
        try
        {
            _logger.LogInformation("Generating test cases for {Language} code", request.Language);

            await Task.Delay(1500); // Simulate generation time

            return new List<GeneratedTestCase>
            {
                new() { Input = "[]", ExpectedOutput = "0", Description = "Empty array test case" },
                new() { Input = "[1]", ExpectedOutput = "1", Description = "Single element test case" },
                new() { Input = "[1, 2, 3, 4, 5]", ExpectedOutput = "15", Description = "Basic functionality test" },
                new() { Input = "[-1, -2, -3]", ExpectedOutput = "-6", Description = "Negative numbers test case" },
                new() { Input = "[1000000]", ExpectedOutput = "1000000", Description = "Large number test case" }
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error generating test cases");
            throw;
        }
    }

    public async Task<PlagiarismResult> CheckPlagiarismAsync(PlagiarismCheckRequest request)
    {
        try
        {
            _logger.LogInformation("Checking plagiarism for question {QuestionId}", request.QuestionId);

            await Task.Delay(2000); // Simulate plagiarism check time

            // Mock plagiarism result
            var random = new Random();
            var similarityScore = random.NextDouble() * 0.3; // 0-30% similarity

            return new PlagiarismResult
            {
                SimilarityScore = similarityScore,
                Matches = similarityScore > 0.15 ? new List<PlagiarismMatch>
                {
                    new()
                    {
                        UserId = Guid.NewGuid().ToString(),
                        Similarity = similarityScore,
                        MatchedLines = new List<int> { 5, 6, 7, 12, 13 }
                    }
                } : new List<PlagiarismMatch>()
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error checking plagiarism");
            throw;
        }
    }

    public async Task<CodeExplanationResult> ExplainCodeAsync(CodeExplanationRequest request)
    {
        try
        {
            _logger.LogInformation("Explaining {Language} code", request.Language);

            await Task.Delay(1000); // Simulate explanation generation

            return new CodeExplanationResult
            {
                Explanation = "This code implements a solution using a systematic approach. It processes the input data, applies the required logic, and returns the expected result.",
                KeyPoints = new List<string>
                {
                    "Uses efficient data structures for optimal performance",
                    "Handles edge cases appropriately",
                    "Follows clean code principles with readable variable names",
                    "Implements the algorithm with proper time complexity"
                },
                AlgorithmUsed = "Two-pointer technique / Hash map approach"
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error explaining code");
            throw;
        }
    }

    private async Task<AIEvaluationResult> GenerateMockEvaluationAsync(CodeEvaluationRequest request)
    {
        await Task.Delay(2000); // Simulate AI processing time

        var random = new Random();
        var score = random.Next(60, 101); // 60-100 score

        return new AIEvaluationResult
        {
            Score = score,
            Feedback = GenerateFeedbackBasedOnScore(score),
            Suggestions = await GetOptimizationSuggestionsAsync(new CodeOptimizationRequest 
            { 
                Code = request.Code, 
                Language = request.Language 
            }),
            CodeQuality = new CodeQualityMetrics
            {
                Readability = random.Next(6, 11),
                Efficiency = random.Next(5, 11),
                Maintainability = random.Next(6, 11),
                BestPractices = random.Next(5, 11)
            },
            TestResults = GenerateMockTestResults(),
            OverallAssessment = score >= 85 ? "Excellent solution!" : 
                              score >= 70 ? "Good solution with room for improvement." : 
                              "Solution needs significant improvements."
        };
    }

    private string GenerateFeedbackBasedOnScore(int score)
    {
        return score switch
        {
            >= 90 => "Outstanding solution! Your code demonstrates excellent problem-solving skills, optimal complexity, and clean implementation.",
            >= 80 => "Very good solution! The logic is correct and efficient. Minor improvements could be made in code organization.",
            >= 70 => "Good solution! The algorithm works correctly but could benefit from optimization and better code structure.",
            >= 60 => "Acceptable solution! The basic logic is correct but needs improvements in efficiency and code quality.",
            _ => "The solution needs significant improvements in both logic and implementation."
        };
    }

    private List<TestResult> GenerateMockTestResults()
    {
        var testCases = new[]
        {
            "Basic functionality test",
            "Edge case: empty input",
            "Edge case: single element",
            "Large input test",
            "Negative numbers test",
            "Boundary values test"
        };

        var random = new Random();
        return testCases.Select(testCase => new TestResult
        {
            TestCase = testCase,
            Passed = random.NextDouble() > 0.2, // 80% pass rate
            Expected = $"Expected result for {testCase}",
            Actual = $"Actual result for {testCase}",
            ExecutionTime = random.Next(10, 150) // 10-150ms
        }).ToList();
    }
}
